### PR TITLE
[Codegen] Add consumer fusion

### PR DESCRIPTION
**Pull request by pashu123**
Adds consumer fusion to tile and distribute using for all pass.

### CODE REVIEW

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! 

Could you also add a test for multiple consumers fused in?

**Comment by pashu123** [comment]
@MaheshRavishankar, as discussed, I've added the case of multiple consumers.
This is failing for consumer fusion with multiple users. I tried this example: https://gist.github.com/pashu123/a0f28b70e393d6841a8d20668400c090. Do you have any suggestions?

**Comment by MaheshRavishankar** [comment]
Looks good. There is an ASAN error in this code. Could you look into this. Ill take your example for multiple uses and see upstream what to do.

**Comment by pashu123** [comment]
> Looks good. There is an ASAN error in this code. Could you look into this. Ill take your example for multiple uses and see upstream what to do.

The consumer fusion has to happen after the scf.forall cleanups. The ASAN error was due to that. I've updated the code.

File: compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp [hunk 1]
```diff
@@ -294,6 +340,7 @@ void TileAndDistributeToWorkgroupsUsingForallOpPass::runOnOperation() {
 
   // If the `tilableOp` is a `memref` op, then just tile the operation.
   SmallVector<LoopLikeOpInterface> tilingLoops;
+  Operation *lastTiledProducerForConsumerFusion = nullptr;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp]: The name is misleading. Just call this `rootTiledOp`  or something. `lastTiledProducer..` sounds like the last producer that is fused in.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**